### PR TITLE
Add logging of calls to run() and validate() methods

### PR DESCRIPTION
**Pull request by xylar**
This should make it easier to find the code that was run in each test case.

### CODE REVIEW

**Comment by xylar** [comment]
## Testing

I ran the `nightly` suite on my Linux laptop with Gnu and MPICH.  I saw:
```
$ grep "Calling:" case_outputs/*
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.decomp_test.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.decomp_test.validate()
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.default.run()
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.run()
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.forward.run()
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.default.validate()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.restart_test.run()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.run()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.run()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.threads_test.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.threads_test.validate()
...
```
as expected.

**Comment by xylar** [comment]
@mark-petersen, this is meant to partially address functionality that I understood you to be requesting in #294.

@matthewhoffman, all I'm asking in a review is for you to say whether you're okay with this change or if you would like to see something else.

**Comment by mark-petersen** [comment]
@xylar thanks, this is exactly what I was looking for. The main usage of this is to see a line like 
```
Calling: compass.ocean.tests.global_ocean.mesh.qu240.QU240Mesh.run()
```
and copy that string to look for that in the compass files. For this example, I would use
```
cd $compass_dir
vi compass/ocean/tests/global_ocean/mesh/qu240/__init__.py 
# look for class QU240Mesh(MeshStep):
```
When I do that, I would need to copy the string from the `Calling: XXX` in the output, and replace all the `.` with `/`. Is there a better way to do that? Would it be appropriate to print the message in the path format, like
```
Calling: QU240Mesh.run() in compass/ocean/tests/global_ocean/mesh/qu240/__init__.py
```
or perhaps better
```
Calling: compass.ocean.tests.global_ocean.mesh.qu240.QU240Mesh.run()
 in compass/ocean/tests/global_ocean/mesh/qu240/__init__.py
```
or similar, so there is a quick string to grab to view that function? Thanks.

I'm guessing that the period format is what python knows internally, and it is just by design that the directories are named the same thing, so python doesn't actually know the path explicitly when it prints this message. That is why you printed the period format, right?

**Comment by xylar** [comment]
I can add the path as well.  It will be challenging to remove the base absolute paths and give relative paths like you have requested, so I think it will need to be a full path.  One reason that python hesitates to give file paths is that files and code are sometime used directly out of compressed archives where you don't have access.  But that would never be true in development mode and probably not when we're using a released conda package either.

I will add this shortly.

**Comment by xylar** [comment]
@mark-petersen, I have added file paths:
```
$ grep -a1 "Calling:" case_outputs/ocean_baroclinic_channel_10km_*
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.decomp_test.DecompTest.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/decomp_test/__init__.py
--
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.InitialState.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/initial_state.py
--
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.Forward.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/forward.py
--
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.Forward.run()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/forward.py
--
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log:Calling: compass.ocean.tests.baroclinic_channel.decomp_test.DecompTest.validate()
case_outputs/ocean_baroclinic_channel_10km_decomp_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/decomp_test/__init__.py
--
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.default.Default.run()
case_outputs/ocean_baroclinic_channel_10km_default.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/default/__init__.py
--
case_outputs/ocean_baroclinic_channel_10km_default.log-
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.InitialState.run()
case_outputs/ocean_baroclinic_channel_10km_default.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/initial_state.py
--
case_outputs/ocean_baroclinic_channel_10km_default.log-
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.forward.Forward.run()
case_outputs/ocean_baroclinic_channel_10km_default.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/forward.py
--
case_outputs/ocean_baroclinic_channel_10km_default.log-
case_outputs/ocean_baroclinic_channel_10km_default.log:Calling: compass.ocean.tests.baroclinic_channel.default.Default.validate()
case_outputs/ocean_baroclinic_channel_10km_default.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/default/__init__.py
--
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.restart_test.RestartTest.run()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/restart_test/__init__.py
--
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.InitialState.run()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/initial_state.py
--
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.Forward.run()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/forward.py
--
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.Forward.run()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/forward.py
--
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-
case_outputs/ocean_baroclinic_channel_10km_restart_test.log:Calling: compass.ocean.tests.baroclinic_channel.restart_test.RestartTest.validate()
case_outputs/ocean_baroclinic_channel_10km_restart_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/restart_test/__init__.py
--
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.threads_test.ThreadsTest.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/threads_test/__init__.py
--
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.initial_state.InitialState.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/initial_state.py
--
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.Forward.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/forward.py
--
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.forward.Forward.run()
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/forward.py
--
case_outputs/ocean_baroclinic_channel_10km_threads_test.log-
case_outputs/ocean_baroclinic_channel_10km_threads_test.log:Calling: compass.ocean.tests.baroclinic_channel.threads_test.ThreadsTest.validate()
```

Please test this out and let me know if it's what you need.

**Comment by xylar** [comment]
@mark-petersen and @matthewhoffman, if you could both test this by running a suite, a test case and a step, that would be great.  The output seemed acceptable to me.  You get this logging info about the test case to the terminal when you run a test case on its own, but that is probably fine since we don't create a log file for the test case right now, just for each step.  So there is nowhere but the terminal for logging info about the test case.

**Review by mark-petersen — APPROVED** [review]
Thanks! I reran the test suite, and this is perfect! That was a missing piece for me, and this will help me navigate compass from the run directories. The full path is actually better, so the user with many branches (like me) does not get confused!

**Comment by xylar** [comment]
Thanks, @mark-petersen, I really appreciate your enthusiasm. This was a very good suggestion.

**Review by matthewhoffman — APPROVED** [review]
This adds information I've also always thought would be useful.  I'm happy to approve as is, but I'm going to throw out one possible suggestions -  Do you want to have these log messages identified slightly more specifically, for example say "COMPASS calling:" or "COMPASS call to:" instead of just "Calling:"?

**Comment by matthewhoffman** [comment]
@trhille , do you have any input on this PR?  It adds these lines saying when the major operations are being called and from where, like:
```
Calling: compass.landice.tests.dome.decomposition_test.DecompositionTest.run()
  in /turquoise/usr/projects/climate/mhoffman/mpas/compass/compass/landice/tests/dome/decomposition_test/__init__.py

  * step: setup_mesh

Calling: compass.landice.tests.dome.setup_mesh.SetupMesh.run()
  in /turquoise/usr/projects/climate/mhoffman/mpas/compass/compass/landice/tests/dome/setup_mesh.py



************************************************************
MPAS_CELL_CULLER:
  C++ version
...
```

**Comment by xylar** [comment]
@matthewhoffman, I made your suggested change.  Could you rerun a quick test?

@trhille, let me know if you'd like to see other modifications.

**Comment by xylar** [comment]
Looking at the code, there's enough redundancy that I think it would be good to generalize this into a little function.  I'll do that quickly and retest.

**Comment by xylar** [comment]
This is a little more complicated than I realized.  Sometimes, `run()` (or `validate()`) isn't defined in the given class but rather it is inherited from a parent class.  It would be better to give the location for the `run()` method accurately because otherwise developers will be looking for the method in the wrong file.  But it is also important to indicate the inheritance in these cases so I need to think about how to do that.

**Comment by trhille** [comment]
I don't need any other modifications, but I'll wait to review until you add the new function, @xylar.

**Comment by xylar** [comment]
I modified things to account for inheritance and to reduce the code redundancy as I mentioned above.  Here is the current behavior for one example step:
```
$ compass run

compass calling: compass.ocean.tests.baroclinic_channel.default.Default.run()
  inherited from: compass.testcase.TestCase.run()
  in /home/xylar/code/compass/log_run_methods/compass/testcase.py


compass calling: compass.ocean.tests.baroclinic_channel.initial_state.InitialState.run()
  in /home/xylar/code/compass/log_run_methods/compass/ocean/tests/baroclinic_channel/initial_state.py

...

compass calling: compass.ocean.tests.baroclinic_channel.default.Default.validate()
  inherited from: compass.testcase.TestCase.validate()
  in /home/xylar/code/compass/log_run_methods/compass/testcase.py
```

**Comment by xylar** [comment]
@matthewhoffman and @trhille, this is ready for you to have another look.  You can either test it yourselves or review by inspection base on my example above.

File: compass/logging.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        The logger to log the method path and file path to
+    """
+    if not inspect.ismethod(method):
+        raise ValueError('The "method" argument must be a method')
+
+    method_name = method.__name__
+    child_class = method.__self__.__class__
+    child_location = f'{child_class.__module__}.{child_class.__name__}'
+    actual_class = None
+    for cls in inspect.getmro(child_class):
```
Comment (trhille) [inline compass/logging.py:27]: Can you add a brief comment explaining this loop?

**Review by trhille — CHANGES_REQUESTED** [review]
@xylar, nice work! I tested this on the Humboldt default test case and got the expected behavior:
```
compass calling: compass.landice.tests.humboldt.default.Default.run()
  inherited from: compass.testcase.TestCase.run()
  in /turquoise/usr/projects/climate/trhille/compass/compass/testcase.py

  * step: mesh

compass calling: compass.landice.tests.humboldt.mesh.Mesh.run()
  in /turquoise/usr/projects/climate/trhille/compass/compass/landice/tests/humboldt/mesh.py

...

compass calling: compass.landice.tests.humboldt.default.Default.validate()
  inherited from: compass.testcase.TestCase.validate()
  in /turquoise/usr/projects/climate/trhille/compass/compass/testcase.py
```
I think adding a comment where indicated would be helpful, but otherwise this looks good!

File: compass/logging.py (hunk 1 above)
Comment (xylar) [reply compass/logging.py:41]: @trhille, let me know if the added comments work for you.

**Comment by xylar** [comment]
I find it mildly problematic (both as a reviewer and as someone whose work is being reviewed) that GitHub chooses to mark requested changes in bright red, just like failed CI tests.  It seems pretty aggressive.

**Comment by matthewhoffman** [comment]
I'm happy with the changes by inspection, and if Trevor already beat me to testing, I'm happy to sign off on that!  Thanks for expanding this functionality so much based on our offhand suggestions!  I really like the way this works now.  I think that will help with debugging a lot and hopefully help new developers learn their way around.

File: compass/logging.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    # expect they always are in compass.  Still we'll check to make sure.
+    try:
+        class_file = inspect.getfile(actual_class)
+    except TypeError:
+        class_file = None
+
+    # log what we found out
+    logger.info(f'compass calling: {child_location}.{method_name}()')
+    if child_location != actual_location:
+        # okay, so we're inheriting this method form somewhere else.  Better
```
Comment (trhille) [inline compass/logging.py:62]: "form" should be "from"

**Review by trhille — APPROVED** [review]
Looks great! I marked one little typo in a comment.

File: compass/logging.py (hunk 2 above)
Comment (xylar) [reply compass/logging.py]: Thanks, fixed.

**Comment by xylar** [comment]
> I think that will help with debugging a lot and hopefully help new developers learn their way around.

I didn't really add anything to the docs (mostly because it would get lost) but developers could, in theory, add `log_method_call()` calls wherever they are helpful for debugging.

**Pull request merged**